kyo-core/shared/src/main/scala/kyo/Clock.scala

```diff
@@ -104,21 +128,158 @@ object Clock:
+        if factor == 1 then v
+        else
+            use { clock =>
+                IO.Unsafe {
+                    val shifted =
+                        new Unsafe:
+                            val underlying  = clock.unsafe
+                            val start       = underlying.now()
+                            val sleepFactor = (1.toDouble / factor)
+                            def now()(using AllowUnsafe) =
+                                val diff = underlying.now() - start
+                                start + (diff * factor)
+                            end now
+                            override def sleep(duration: Duration) =
+                                underlying.sleep(duration * sleepFactor)
+                    let(Clock(shifted))(v)
+                }
+            }
+    end withTimeShift
+
+    /** Interface for controlling time in a test environment.
+      *
+      * WARNING: TimeControl is not thread-safe. All operations should be performed sequentially to avoid race conditions and unexpected
+      * behavior.
+      */
+    trait TimeControl:
+        /** Sets the current time to a specific instant.
+          *
+          * @param now
+          *   The instant to set the current time to
+          * @return
+          *   Unit effect that updates the current time
+          */
+        def set(now: Instant): Unit < IO
+
+        /** Advances the current time by the specified duration.
+          *
+          * @param duration
+          *   The duration to advance time by
+          * @return
+          *   Unit effect that advances the current time
+          */
+        def advance(duration: Duration): Unit < IO
+    end TimeControl
+
+    /** Runs an effect with a controlled Clock that allows manual time manipulation. This is primarily intended for testing scenarios where
+      * precise control over time progression is needed.
+      *
+      * Note: TimeControl is not thread-safe. Operations on TimeControl should be performed sequentially within the same fiber.
+      *
+      * @param f
+      *   A function that takes a TimeControl and returns an effect
+      * @return
+      *   The result of running the effect with controlled time
+      */
+    def withTimeControl[A, S](f: TimeControl => A < S)(using Frame): A < (IO & S) =
+        IO.Unsafe {
+            val controlled =
+                new Unsafe with TimeControl:
+                    @volatile var current = Instant.Epoch
+
+                    case class Task(deadline: Instant) extends IOPromise[Nothing, Unit] with Delayed:
+                        def getDelay(unit: TimeUnit): Long =
+                            (deadline - current).to(unit)
+                        def compareTo(other: Delayed): Int =
+                            deadline.toJava.compareTo(other.asInstanceOf[Task].deadline.toJava)
+                    end Task
+
+                    val queue = new DelayQueue[Task]
```

Comment:
As simple timer implementation using `DelayQueue`. The queue is `synchronized` so `sleep` can be called concurrently by multiple fibers

Reply:
I had to change it to a manually-synchronized `PriorityQueue` because `DelayQueue` isn't available in js